### PR TITLE
Improve the performance of LP file parsing

### PR DESCRIPTION
**Pull request by arcondello**
Adds a C++ LP file reader from HiGHS (https://github.com/ERGO-Code/HiGHS/tree/501b30a) and wraps it.

So far just a DRAFT to see what CI makes of this. There is a bunch of docs/cleanup still to be done.

Closes https://github.com/dwavesystems/dimod/issues/928

### CODE REVIEW

File: tests/test_serialization_lp.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+        x, y = dimod.Reals('xy')
+
+        cqm = read_lp(lp)
+
+        self.assertFalse(cqm.constraints)
+        self.assertTrue(cqm.objective.is_equal(x + y))
+
+    # todo
+    # def test_scientific_notation(self):
```
Comment (randomir) [inline tests/test_serialization_lp.py:102]: :tada: (closes #1132)

**Comment by randomir** [comment]
How about you contribute this as a PR to https://github.com/scipy/scipy/issues/14593 instead? SciPy seems to [already include HiGHS](https://github.com/scipy/scipy/pull/15217).

**Comment by arcondello** [comment]
> How about you contribute this as a PR to [scipy/scipy#14593](https://github.com/scipy/scipy/issues/14593) instead? SciPy seems to [already include HiGHS](https://github.com/scipy/scipy/pull/15217).

I thought about it, but a few reasons to not:
1) This way we don't make any intermediate objects, which is important for large models
2) To access it from the SciPy namespace, HiGHS would need to partly restructure their code (possible can be gotten around)
3) There are a few bug fixes that would also need to be made to HiGHs. Once https://github.com/ERGO-Code/HiGHS/issues/840, https://github.com/ERGO-Code/HiGHS/pull/841, and https://github.com/ERGO-Code/HiGHS/pull/839 are resolved we could proceed
5) Would need HiGHS to also support scientific notation
6) Would need SciPy to then use the latest HiGHS
7) We would need to add SciPy as a dependency to dimod, as well as wait for it to deploy. That would also have downstream consequences on the range of NumPy versions we can support

That said, only (1) is a long-term problem. Once (3) and (5) are resolved, I'll hopefully have time to put together a PR

**Comment by arcondello** [comment]
There is a bit more test restructure that could be done, but I am eager to release 0.11 and want to get this in. In addition to the current tests, I tested it on the entirety of [qplib](https://qplib.zib.de/)

File: releasenotes/notes/feature-c++-lp-reader-1d5109d86ce778ed.yaml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  - Add ``read_lp_file()`` and ``read_lp()`` functions.
+  - Improve the performance of the ``ConstrainedQuadraticModel.from_lp_file()`` method.
+upgrade:
+  - |
+    Some malformed LP-files that previously caused ``ConstrainedQuadraticModel.from_lp_file()``
+    to emit a warning now raise an error instead.
+deprecations:
+  - |
+    Deprecate ``ConstrainedQuadraticModel.from_lp_file()`` method. Users should
+    use the more explicit ``read_lp_file()`` and ``read_lp()`` functions instead.
```
Comment (randomir) [inline releasenotes/notes/feature-c++-lp-reader-1d5109d86ce778ed.yaml:12]: Shouldn't users now use `dimod.lp.load` and `dimod.lp.loads`?

Comment (randomir) [reply releasenotes/notes/feature-c++-lp-reader-1d5109d86ce778ed.yaml:12]: Oh, I see you fixed it in a follow-up commit.

**Comment by lesshaste** [comment]
> > How about you contribute this as a PR to [scipy/scipy#14593](https://github.com/scipy/scipy/issues/14593) instead? SciPy seems to [already include HiGHS](https://github.com/scipy/scipy/pull/15217).
> 
> I thought about it, but a few reasons to not:
> 
> 1. This way we don't make any intermediate objects, which is important for large models
> 2. To access it from the SciPy namespace, HiGHS would need to partly restructure their code (maybe)
> 3. There are a few bug fixes that would also need to be made to HiGHs. Once [Trying to load an LP file without a newline at the end of the file leads to out-of-memory errors ERGO-Code/HiGHS#840](https://github.com/ERGO-Code/HiGHS/issues/840), [Fix EOF detection in LP file reader ERGO-Code/HiGHS#841](https://github.com/ERGO-Code/HiGHS/pull/841), and [Allow LP file parse to handle var*var in expressions ERGO-Code/HiGHS#839](https://github.com/ERGO-Code/HiGHS/pull/839) are resolved we could proceed
> 4. Would need HiGHS to also support scientific notation
> 5. Would need SciPy to then use the latest HiGHS
> 6. We would need to add SciPy as a dependency to dimod, as well as wait for it to deploy. That would also have downstream consequences on the range of NumPy versions we can support
> 
> That said, only (1) is a long-term problem. Once (3), (4), and (5) are resolved, I'll hopefully have time to put together a PR

It looks like 3 and 5 at least have happened now.

**Pull request merged**